### PR TITLE
HC control of default boundary condition - part one

### PR DESCRIPTION
**Pull request by jbathmann**
This is the first part of PR #2466 containing only the changes in process assembly and test updates.

The HC-Test files are updated, such that they control the functionality of the new implementation. Hereby, the elder and the goswami benchmark agree with experimental observations. Tests for the non-advective FEM type  will be added with PR #2466.

### CODE REVIEW

File: ProcessLib/ComponentTransport/CreateComponentTransportProcess.cpp [hunk 1]
```diff
@@ -148,6 +148,10 @@ std::unique_ptr<Process> createComponentTransportProcess(
         std::copy_n(b.data(), b.size(), specific_body_force.data());
     }
 
+    bool const non_advective_form =
+        //! \ogs_file_param{prj__processes__process__ComponentTransport__specific_body_force}
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/CreateComponentTransportProcess.cpp]: The `\ogs_file_param` comment does not match:
```suggestion
        //! \ogs_file_param{prj__processes__process__ComponentTransport__non_advective_form}
```

Comment (jbathmann) [reply ProcessLib/ComponentTransport/CreateComponentTransportProcess.cpp]: changed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 2]
```diff
@@ -341,35 +344,41 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+            {
+                BC.noalias() -=
+                    N.transpose() *
+                    (mass_density_flow.dot(dNdx * C_nodal_values) * w);
+            }
+            MCC.noalias() += N_t_N * (R_times_phi * density * w);
+            KCC.noalias() += dNdx.transpose() * hydrodynamic_dispersion * dNdx *
+                                 (density * w) +
+                             N_t_N * (decay_rate * porosity *
+                                      retardation_factor * density * w);
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: You can reuse `R_times_phi` (` = porosity*retardation_factor`) here.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 3]
```diff
@@ -341,35 +344,41 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+            MpC.noalias() += N_t_N * (porosity * drho_dC * w);
 
             // Calculate Mpp, Kpp, and bp in the first loop over components
             if (component_id == 0)
             {
-                Mpp.noalias() += w * N.transpose() * porosity * drho_dp * N;
+                Mpp.noalias() += N_t_N * (porosity * drho_dp * w);
                 Kpp.noalias() +=
-                    w * dNdx.transpose() * density * K_over_mu * dNdx;
+                    dNdx.transpose() * (density * K_over_mu * w) * dNdx;
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: `K_over_mu` can be a matrix, too.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 4]
```diff
@@ -341,35 +344,41 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
 
                 if (_process_data.has_gravity)
                 {
-                    Bp += w * density * density * dNdx.transpose() * K_over_mu *
-                          b;
+                    Bp.noalias() += dNdx.transpose() *
+                                    (density * density * K_over_mu * b * w);
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: If you want to optimize for performance you could move the `density * density` to `w` (`K_over_mu * b` is a vector).

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 5]
```diff
@@ -628,28 +630,49 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+                        vars,
+                        MaterialLib::Fluid::PropertyVariableType::C);
+                local_M.noalias() +=
+                    N_t_N * (R_times_phi * C_int_pt * drho_dC * w);
+            }
+            else
+            {
+                local_b.noalias() -=
+                    N.transpose() *
+                    (density * velocity.dot(dNdx * local_C) * w);
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Could be reformulated as: `velocity.dot(dNdx * local_C) * (density * w)`.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 2 above)
Comment (jbathmann) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: changed

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 3 above)
Comment (jbathmann) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: I moved K_over_mu

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 4 above)
Comment (jbathmann) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: thanks - moved K_over_mu * b

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 5 above)
Comment (jbathmann) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: done

File: Tests/Data/Parabolic/ComponentTransport/VariableNeumannBoundary/vdbc_input.prj [hunk 6]
```diff
@@ -246,15 +246,15 @@
             <components>1</components>
             <order>1</order>
             <initial_condition>c_ini</initial_condition>
-            <boundary_conditions>
+            <boundary_conditions><!--
```
Comment (endJunction) [inline Tests/Data/Parabolic/ComponentTransport/VariableNeumannBoundary/vdbc_input.prj]: If the commented section is important, add a comment in there, remove otherwise.

Comment (jbathmann) [reply Tests/Data/Parabolic/ComponentTransport/VariableNeumannBoundary/vdbc_input.prj]: Comment added.

**Comment by renchao-lu** [comment]
1) Based upon my understanding, conventional and revised formulations have difference only in the advection term. Thus, no changes will be made on other matrices except the local matrix KCC.
2) In my opinion, it is not needed to drop the existing great feature, i.e., dependency of fluid density upon concentration and pressure, in implementing conventional formulation.

For your reference, my expected changes would be like this (as mentioned in my comment in PR #2409 ):
if (use conventional formulation) // (default case)
{
KCC.noalias() +=

        (N.transpose() * velocity.transpose() * density * dNdx +

        (-dNdx.transpose() * velocity * density * N +

        dNdx.transpose() * density * hydrodynamic_dispersion * dNdx +
        N.transpose() * decay_rate * porosity * retardation_factor *
               density * N) *
       w;
}
else // use revised formulation (non-default case)
{

KCC.noalias() +=

        (-dNdx.transpose() * velocity * density * N +

        dNdx.transpose() * density * hydrodynamic_dispersion * dNdx +
        N.transpose() * decay_rate * porosity * retardation_factor *
               density * N) *
       w;
}

**Comment by jbathmann** [comment]
>     1. Based upon my understanding, conventional and revised formulations have difference only in the advection term. Thus, no changes will be made on other matrices except the local matrix KCC.
> 
>     2. In my opinion, it is not needed to drop the existing great feature, i.e., dependency of fluid density upon concentration and pressure, in implementing conventional formulation.
> 
> 
> For your reference, my expected changes would be like this (as mentioned in my comment in PR #2409 ):
> 
> ```
> if (use conventional formulation) // (default case)
> {
> KCC.noalias() +=
> 
>         (N.transpose() * velocity.transpose() * density * dNdx +
> 
>         (-dNdx.transpose() * velocity * density * N +
> 
>         dNdx.transpose() * density * hydrodynamic_dispersion * dNdx +
>         N.transpose() * decay_rate * porosity * retardation_factor *
>                density * N) *
>        w;
> }
> else // use revised formulation (non-default case)
> {
> 
> KCC.noalias() +=
> 
>         (-dNdx.transpose() * velocity * density * N +
> 
>         dNdx.transpose() * density * hydrodynamic_dispersion * dNdx +
>         N.transpose() * decay_rate * porosity * retardation_factor *
>                density * N) *
>        w;
> }
> ```

@renchao-lu Thanks for your comments.
Some thoughts on this: 

1. If one uses the advective form (non-default case in your comment; default case in this PR), the coupling matrix entries for M^{cc} and M^{cp} do change as well. Thus their assembly needs to be changed accordingly. Maybe you might want to have a short look on equations (2.17), (2.24) and (2.25) in the updated derivation included in this PR, for further details.
2. Thanks for the remark on "-dNdx.transpose() * velocity * density * N", which I currently add to the vector b. I will change this!
3. The "dependency of fluid density upon concentration and pressure" are not dropped in this PR. The only change is the choice of the default boundary conditions.

**Comment by renchao-lu** [comment]
Thanks for your patient interpretation in the derivation of weak formulation. Now I understand your concerns. @jbathmann

**Comment by endJunction** [comment]
Now, if all parties are happy, let's merge it.

**Comment by renchao-lu** [comment]
It is fine to me. @endJunction
